dvc/repo/params/show.py

```diff
@@ -117,74 +95,101 @@ def _collect_vars(repo, params, stages=None) -> Dict:
+    on_error: str = "return",
+):
+    assert on_error in ("raise", "return", "ignore")
+
+    files_keypaths = _collect_params(
+        repo,
+        targets=targets,
+        stages=stages,
+        deps_only=deps_only,
+        default_file=ParamsDependency.DEFAULT_PARAMS_FILE,
+    )
+
+    data: Dict[str, FileResult] = {}
+
+    dvcfs = repo.dvcfs
+    for fs_path, result in _read_params(dvcfs, files_keypaths):
+        repo_path = fs_path.lstrip(dvcfs.root_marker)
+        repo_os_path = os.sep.join(repo.fs.path.parts(repo_path))
+        if not isinstance(result, Exception):
+            data.update({repo_os_path: FileResult(data=result)})
+            continue
+
+        if on_error == "raise":
+            raise result
+        if on_error == "return":
+            data.update({repo_os_path: FileResult(error=result)})
+
+    # vars_params = _collect_vars(repo, params, stages=stages)
+
+    # NOTE: only those that are not added as a ParamDependency are
+    # included so we don't need to recursively merge them yet.
+    # for key, vals in vars_params.items():
+    #     params[key]["data"] = vals
+    return data
+
+
 @locked
 def show(
-    repo,
-    revs=None,
-    targets=None,
-    deps=False,
-    onerror: Optional[Callable] = None,
-    stages=None,
-    hide_workspace=True,
-):
-    if onerror is None:
-        onerror = onerror_collect
+    repo: "Repo",
+    targets: Optional[List[str]] = None,
+    stages: Optional[List[str]] = None,
+    deps_only: bool = False,
+    all_branches: bool = False,
+    all_tags: bool = False,
+    revs: Optional[List[str]] = None,
+    all_commits: bool = False,
+    hide_workspace: bool = True,
+    on_error: str = "return",
+) -> Dict[str, Result]:
+    assert on_error in ("raise", "return", "ignore")
     res = {}
 
     targets = ensure_list(targets)
     targets = [repo.dvcfs.from_os_path(target) for target in targets]
 
-    for branch in repo.brancher(revs=revs):
-        params = error_handler(_gather_params)(
-            repo=repo,
-            targets=targets,
-            deps=deps,
-            onerror=onerror,
-            stages=stages,
-        )
-
-        if params:
-            res[branch] = params
+    for rev in repo.brancher(
+        revs=revs,
+        all_branches=all_branches,
+        all_tags=all_tags,
+        all_commits=all_commits,
+    ):
+        try:
+            params = _gather_params(
+                repo=repo,
+                targets=targets,
+                stages=stages,
+                deps_only=deps_only,
+                on_error=on_error,
+            )
+            res[rev] = Result(data=params)
+        except Exception as exc:  # noqa: BLE001 # pylint:disable=broad-exception-caught
+            if on_error == "raise":
+                raise
+            logger.warning("failed to load params in revision %r, %s", rev, str(exc))
+            if on_error == "return":
+                res[rev] = Result(error=exc)
```

Comment:
I am moving away from `onerror` callback. Moving forward, I'd prefer to have a tighter control, and have explicit places (and fewer places) where we ignore/suppress exception.

I am okay with `onerror` callback as a way to decide whether or not to ignore an exception as an exception handler, but I don't like it as a way to change return type.

Reply:
Should we expect anything to fail that wouldn't fail before?

Reply:
Nothing should fail. We are still suppressing those.